app/modules/pwm2.c

```diff
@@ -0,0 +1,553 @@
+  uint32_t currentInterruptCounter;
+  uint16_t gpioMask;
+} pwm2_pin_interrupt_t;
+
+typedef struct
+{
+  pwm2_pin_interrupt_t pin[GPIO_PIN_NUM];
+  uint16_t enabledGpioMask;
+} pwm2_interrupt_handler_data_t;
+
+typedef struct
+{
+  uint32_t pulseResolutions;
+  uint32_t divisableFrequency;
+  uint32_t frequencyDivisor;
+  uint32_t duty;
+  uint32_t resolutionCPUTicks;
+  uint32_t resolutionInterruptCounterMultiplier;
+} pwm2_pin_setup_t;
+
+typedef struct
+{
+  pwm2_pin_setup_t pin[GPIO_PIN_NUM];
+  uint32_t interruptTimerCPUTicks;
+  uint32_t interruptTimerTicks;
+  bool isStarted;
+} pwm2_setup_data_t;
+
+typedef struct
+{
+  pwm2_interrupt_handler_data_t interruptData;
+  pwm2_setup_data_t setupData;
+} pwm2_module_data_t;
+
+//############################
+// module vars, lazy initialized, allocated only if pwm2 is being used
+
+static pwm2_module_data_t *moduleData = NULL;
+
+//############################
+
+static bool isPinSetup(const pwm2_module_data_t *data, const uint8_t pin)
+{
+  return data->setupData.pin[pin].pulseResolutions > 0;
+}
+
+static uint32_t getCPUTicksPerSec()
+{
+  return system_get_cpu_freq() * 1000000;
+}
+
+static uint8_t getCpuTimerTicksDivisor()
+{
+  return system_get_cpu_freq() == 80 ? PWM2_TMR_MAGIC_80MHZ : PWM2_TMR_MAGIC_160MHZ;
+}
+
+static uint32_t findGCD(uint32_t n1, uint32_t n2)
+{
+  while (n1 != n2)
+  {
+    if (n1 > n2)
+      n1 -= n2;
+    else
+      n2 -= n1;
+  }
+  return n1;
+}
+
+static uint32_t findGreatesCommonDividerForTimerTicks(uint32_t newTimerTicks, uint32_t oldTimerTicks)
+{
+  return oldTimerTicks == 0 ? newTimerTicks : findGCD(newTimerTicks, oldTimerTicks);
+}
+
+static uint16_t findAllEnabledGpioMask(pwm2_module_data_t *moduleData)
+{
+  uint16_t enableGpioMask = 0;
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (moduleData->setupData.pin[i].pulseResolutions > 0)
+    {
+      enableGpioMask |= moduleData->interruptData.pin[i].gpioMask;
+    }
+  }
+  return enableGpioMask;
+}
+
+static uint32_t findCommonCPUTicksDivisor(pwm2_module_data_t *moduleData)
+{
+  uint32_t gcdCPUTicks = 0;
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (moduleData->setupData.pin[i].pulseResolutions > 0)
+    {
+      gcdCPUTicks = findGreatesCommonDividerForTimerTicks(moduleData->setupData.pin[i].resolutionCPUTicks, gcdCPUTicks);
+    }
+  }
+  return gcdCPUTicks;
+}
+
+static uint32_t cpuToTimerTicks(uint32_t cpuTicks)
+{
+  return cpuTicks / getCpuTimerTicksDivisor();
+}
+
+static void updatePinResolutionToInterruptsMultiplier(pwm2_pin_setup_t *sPin, uint32_t timerCPUTicks)
+{
+  sPin->resolutionInterruptCounterMultiplier = sPin->resolutionCPUTicks / timerCPUTicks;
+}
+
+static void updatePinPulseToInterruptsCounter(pwm2_pin_interrupt_t *iPin, pwm2_pin_setup_t *sPin)
+{
+  iPin->pulseInterruptCcounter = (sPin->pulseResolutions + 1) * sPin->resolutionInterruptCounterMultiplier;
+}
+
+static uint8_t getDutyAdjustment(const uint32_t duty, const uint32_t pulse)
```

Comment:
Conditional operator considered unhealthy :wink:

Reply:
i'm not catching this up, what do you mean?